compiler/rustc_middle/src/ty/error.rs

```diff
@@ -771,11 +771,40 @@ fn foo(&self) -> Self::T { String::new() }
     ) -> bool {
         let assoc = self.associated_item(proj_ty.item_def_id);
         if let ty::Opaque(def_id, _) = *proj_ty.self_ty().kind() {
+            let opaque_local_def_id = def_id.expect_local();
+            let opaque_hir_id = self.hir().local_def_id_to_hir_id(opaque_local_def_id);
+            let opaque_hir_ty = match &self.hir().expect_item(opaque_hir_id).kind {
+                hir::ItemKind::OpaqueTy(opaque_hir_ty) => opaque_hir_ty,
+                _ => bug!("The HirId comes from a `ty::Opaque`"),
+            };
+
+            let (assoc_trait_ref, assoc_ty_substs) = proj_ty.trait_ref_and_own_substs(self);
+
+            // If the opaque trait is something like `impl Foo + Bar`
+            // then try to select the span of the right trait.
+            // This may fail if the associated type is defined on a supertrait of
+            // those present in the opaque type definiton: in that case fall back to the
+            // previous (sometimes wrong) span.
+            let span = opaque_hir_ty
+                .bounds
+                .iter()
+                .find_map(|bound| {
+                    let trait_ref_def_id = bound
+                        .trait_ref()?
+                        .trait_def_id()
+                        .expect("All `hir::TraitDef`s should already be resolved");
+                    if trait_ref_def_id != assoc_trait_ref.def_id {
+                        return None;
+                    }
+                    Some(bound.span())
+                })
+                .unwrap_or_else(|| self.def_span(def_id));
```

Comment:
I think we should not produce any suggestion in this case. We could still emit a help message at the def_span talking about the fact that the bound needs to be added here

Reply:
The unfortunate thing about doing that is that we lose suggestions for cases like `impl SubTrait`. I might have had an idea to handle those cases though.